src/wasm/wasm-validator.cpp

```diff
@@ -2874,7 +2924,12 @@ static void validateTables(Module& module, ValidationInfo& info) {
         info.shouldBeTrue(
           expr->is<RefFunc>() || expr->is<RefNull>(),
```

Comment:
One missing part seems to be here, where we validate allowed instructions for elem expressions. My understanding is that we should allow things like `struct.new` and `array.new`. Is that correct? @kripken @tlively

Reply:
I think the answer is no. Only rtts and global.get: https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#constant-expressions

Reply:
@kripken hmm thanks! So far I've been avoiding RTTs to be honest :) so my knowledge of them in GC is limited, and I will dig deeper for sure. But just to be clear, that would mean I can't put an array on an element segment and initialize a table with it, right? E.g.

```
(module
  (type $vec (array f32))
  (table 1 dataref)
  (elem (i32.const 0) dataref (array.new_default $vec (i32.const 3) (rtt.canon $vec)))
)
```

Reply:
So it appears that there's an open issue for extending the list of constant instructions, WebAssembly/gc#189. That answers my questions. I'll make some changes accordingly.

Reply:
@kripken GC's MVP document [says](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#reference-types)

    (rtt <n>? $t) == (ref (rtt <n>? $t))

But we don't seem to parse `(ref (rtt <n>? $t))` at all, right? And we don't consider `(rtt <n>? $t)` to be a reference type? So basically I need to check that the table type is either `.isRef()` or `.isRtt()`.

Reply:
Making RTTs into reference types is a very recent change to the GC spec, that I don't think anyone has implemented. I don't even think it has been discussed at length in the CG actually.

For now, binaryen implements the [GC prototype 3 doc](https://docs.google.com/document/d/1DklC3qVuOdLHSXB5UXghM_syCh-4cMinQ50ICiXnK3Q/edit) which is a snapshot of the spec from earlier (snapshots allow us to interop with compilers and VMs, so we all do the same thing, ignoring the daily churn in the spec). So for us it is correct to not parse `(ref (rtt ..))`, for now.

Reply:
Thanks for the explanation. In that case, I will skip RTT altogether since we don't consider them as reference types and even then, they don't seem to be nullable in the current version of the spec.